src/io.rs

```diff
@@ -84,14 +384,95 @@ mod tests {
-        let mut expected = vec![0; 0];
-        af.read_to_end(&mut actual).unwrap();
-        ef.read_to_end(&mut expected).unwrap();
+            assert_eq!(actual, expected);
+        }
 
-        assert_eq!(actual, expected);
+        let _ = std::fs::remove_file(actual_filename);
+    }
+    #[test]
+    fn peek_next() {
+        use std::{io::Cursor, rc::Rc};
+        let mut buf = Vec::new();
+        let mut rc = Rc::new(&mut buf);
+        let records = vec!["foo", "barr", "baz"];
+        {
+            let mut writer = RecordWriter::new(Rc::get_mut(&mut rc).unwrap());
+            for rec in records.iter() {
+                writer.write_record(rec.as_bytes()).unwrap();
+            }
+        }
+        let read = std::io::BufReader::new(Cursor::new(buf));
+        let mut reader = RecordReader::new(read);
+        let mut ary = [0u8; 4];
+        let mut i = 0;
+        loop {
+            if i < 3 {
+                assert_eq!(
+                    reader.peek_next_len().unwrap().unwrap() as usize,
+                    records[i].len()
+                );
+            }
+            if i == 3 {
+                assert!(reader.peek_next_len().unwrap().is_none());
+            }
+
+            let next = reader.read_next(&mut ary);
+            match next {
+                Ok(res) => match res {
+                    Some(len) => assert_eq!(&ary[0..len], records[i].as_bytes()),
+                    None => break,
+                },
+                Err(RecordReadError::CorruptFile) | Err(RecordReadError::IoError { .. }) => {
+                    break;
+                }
+                _ => {}
+            }
+            i += 1;
+        }
+    }
+
+    #[test]
+    fn seek_over_too_long() {
+        use std::{io::Cursor, rc::Rc};
+        let mut buf = Vec::new();
+        let mut rc = Rc::new(&mut buf);
+        let records = vec!["foo", "barr", "baz"];
+        {
+            let mut writer = RecordWriter::new(Rc::get_mut(&mut rc).unwrap());
+            for rec in records.iter() {
+                writer.write_record(rec.as_bytes()).unwrap();
+            }
+        }
+        let read = std::io::BufReader::new(Cursor::new(buf));
+        let mut reader = RecordReader::new(read);
+        let mut ary = [0u8; 3];
+        let mut i = 0;
+        loop {
+            let next = reader.read_next(&mut ary);
+            if i == 0 {
+                assert_eq!(next.unwrap().unwrap(), 3);
+                assert_eq!(&ary, records[i].as_bytes());
+            } else if i == 1 {
+                let buffer_too_short = match next {
+                    Err(RecordReadError::BufferTooShort { needed: v }) => v == 4,
+                    _ => false,
+                };
+                assert!(buffer_too_short);
+            } else if i == 2 {
+                assert_eq!(next.unwrap().unwrap(), 3);
+                assert_eq!(&ary, records[i].as_bytes());
+            } else {
+                break;
+            }
```

Comment:
Rather than a loop with an if-else that does something different for each index, it would make more sense to unroll it.